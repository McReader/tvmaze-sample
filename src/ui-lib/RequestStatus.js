import PropTypes from "prop-types";
import { CircularProgress } from "@material-ui/core";
import ErrorIcon from '@material-ui/icons/Error';

export const REQUEST_STATUS_IDLE = "idle";
export const REQUEST_STATUS_PENDING = "pending";
export const REQUEST_STATUS_FULFILLED = "fulfilled";
export const REQUEST_STATUS_REJECTED = "rejected";

/**
 * Renders progress bar when "requestStatus" is "idle" or "pending"
 * Renders caution symbol when "requestStatus" is "rejected"
 * Uses "render prop" pattern to render children when "request status" is "fulfilled"
 */
export function RequestStatus({ children, requestStatus = REQUEST_STATUS_IDLE }) {
    if (requestStatus === REQUEST_STATUS_IDLE || requestStatus === REQUEST_STATUS_PENDING) {
        return <CircularProgress  />;
    }

    if (requestStatus === REQUEST_STATUS_REJECTED) {
        return <ErrorIcon />;
    }

    return children();
}

export const requestStatusPropType = PropTypes.oneOf([
    REQUEST_STATUS_IDLE,
    REQUEST_STATUS_PENDING,
    REQUEST_STATUS_FULFILLED,
    REQUEST_STATUS_REJECTED
]);

RequestStatus.propTypes = {
    children: PropTypes.func.isRequired,
    requestStatus: requestStatusPropType,
};
