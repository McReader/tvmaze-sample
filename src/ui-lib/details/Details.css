.Details {
    height: 100%;
    width: 100%;
}

.Details h1 {
    padding: 0 16px;
}

.DetailsSummaryContainer {
    display: flex;
    margin: 0 8px;
}

.DetailsSummary {
    margin-left: 16px;
}

.DetailsCoverImage {
    max-width: 500px;
    min-width: 200px;
}

@media only screen and (max-width: 500px) {
    .DetailsSummaryContainer {
        flex-flow: column;
    }

    .DetailsSummaryContainer {
        margin: 0;
    }

    .DetailsSummary {
        margin-left: 0;
        padding: 0 8px;
    }
}
