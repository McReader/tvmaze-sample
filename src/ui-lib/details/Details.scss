@import "../../design-tokens/variables";

.Details {
	height: 100%;
	width: 100%;
}

.Details h1 {
	padding: $token-spacing-medium;
}

.DetailsSummaryContainer {
	display: flex;
	margin: 0 $token-spacing-small;
}

.DetailsSummary {
	margin-left: $token-spacing-medium;
}

.DetailsCoverImage {
	max-width: 500px;
	min-width: 200px;
}

@media only screen and (max-width: $token-breakpoint-mobile) {
	.DetailsSummaryContainer {
		flex-flow: column;
	}

	.DetailsSummaryContainer {
		margin: 0;
	}

	.DetailsSummary {
		margin-left: 0;
		padding: 0 8px;
	}
}
