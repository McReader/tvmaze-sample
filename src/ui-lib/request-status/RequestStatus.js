import PropTypes from "prop-types"
import ErrorIcon from "@material-ui/icons/Error"

import { ProgressPlaceholder } from "../progress-placeholder/ProgressPlaceholder"
import { FillParentPlaceholder } from "../fill-parent-placeholder/FillParentPlaceholder"

export const REQUEST_STATUS_IDLE = "idle"
export const REQUEST_STATUS_PENDING = "pending"
export const REQUEST_STATUS_FULFILLED = "fulfilled"
export const REQUEST_STATUS_REJECTED = "rejected"

/**
 * Renders progress bar when "requestStatus" is "idle" or "pending"
 * Renders caution symbol when "requestStatus" is "rejected"
 * Uses "render prop" pattern to render children when "request status" is "fulfilled"
 */
export function RequestStatus({
	children,
	requestStatus = REQUEST_STATUS_IDLE,
}) {
	if (
		requestStatus === REQUEST_STATUS_IDLE ||
		requestStatus === REQUEST_STATUS_PENDING
	) {
		return <ProgressPlaceholder />
	}

	if (requestStatus === REQUEST_STATUS_REJECTED) {
		return (
			<FillParentPlaceholder>
				<ErrorIcon />
			</FillParentPlaceholder>
		)
	}

	return children()
}

export const requestStatusPropType = PropTypes.oneOf([
	REQUEST_STATUS_IDLE,
	REQUEST_STATUS_PENDING,
	REQUEST_STATUS_FULFILLED,
	REQUEST_STATUS_REJECTED,
])

RequestStatus.propTypes = {
	children: PropTypes.func.isRequired,
	requestStatus: requestStatusPropType,
}
