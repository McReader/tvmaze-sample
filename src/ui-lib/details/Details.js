import PropTypes from "prop-types";

import { RequestStatus, requestStatusPropType } from "../request-status/RequestStatus";

import { CoverImage, coverImagePropType } from "./cover-image/CoverImage";

import "./Details.css";

export function Details({ requestStatus, details }) {
    return (
        <div className="Details">
            <RequestStatus requestStatus={requestStatus}>
                {() => {
                    const { image, name, summary } = details;

                    return (
                        <>
                            <h1>{name}</h1>
                            <div className="DetailsSummaryContainer">
                                <CoverImage className="DetailsCoverImage" image={image} />
                                <summary
                                    className="DetailsSummary"
                                    dangerouslySetInnerHTML={{ __html: summary }}
                                />
                            </div>
                        </>
                    );
                }}
            </RequestStatus>
        </div>
    );
}

Details.propTypes = {
    details: PropTypes.shape({
        image: coverImagePropType,
        name: PropTypes.string.isRequired,
        summary: PropTypes.string,
    }),
    requestStatus: requestStatusPropType,
};
