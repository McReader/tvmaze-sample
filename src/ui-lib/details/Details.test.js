import renderer from "react-test-renderer"

import { Details } from "./Details"

import show from "../../tvmaze-sdk/shows/showMainInformationSample.json"

import { REQUEST_STATUS_FULFILLED } from "../request-status/RequestStatus"

let tree
let props

beforeEach(() => {
	tree = renderer.create(<Details {...props} />).toJSON()
})

describe(`when request status is "${REQUEST_STATUS_FULFILLED}" and there is a show`, () => {
	beforeAll(() => {
		props = {
			requestStatus: REQUEST_STATUS_FULFILLED,
			details: show,
		}
	})

	test("should match snapshot", () => {
		expect(tree).toMatchSnapshot()
	})
})
