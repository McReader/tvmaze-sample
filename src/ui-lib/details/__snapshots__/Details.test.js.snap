// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when request status is "fulfilled" and there is a show should match snapshot 1`] = `
<div
  className="Details"
>
  <h1>
    Under the Dome
  </h1>
  <div
    className="DetailsSummaryContainer"
  >
    <picture
      className="DetailsCoverImage"
    >
      <source
        media="(max-width:480px"
        srcSet="http://static.tvmaze.com/uploads/images/original_untouched/81/202627.jpg"
      />
      <source
        media="(max-width:768px)"
        srcSet="http://static.tvmaze.com/uploads/images/medium_portrait/81/202627.jpg"
      />
      <img
        alt="Cover"
        className="CoverImageImg"
        src="http://static.tvmaze.com/uploads/images/original_untouched/81/202627.jpg"
      />
    </picture>
    <summary
      className="DetailsSummary"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p><b>Under the Dome</b> is the story of a small town that is suddenly and inexplicably sealed off from the rest of the world by an enormous transparent dome. The town's inhabitants must deal with surviving the post-apocalyptic conditions while searching for answers about the dome, where it came from and if and when it will go away.</p>",
        }
      }
    />
  </div>
</div>
`;
